feat: 별점 등록 기능 구현

jo0oy opened the pull request:
### 💡 Motivation
- 별점 등록 기능을 개발합니다.
- Test 환경 Redis를 위해 Testcontainers Redis를 설정합니다.

### 📌 Changes
- 별점 등록 Request/Response DTO 추가
- 별점 등록 비즈니스 로직 구현, API 구현
- 별점 등록 기능 관련 Test Code 작성 후 동작 확인
- appliacaion.yml 'test' 프로파일에 `spring.jpa.show-sql=true` 설정 추가
- 별점 등록 기능 Rest Docs를 위한 Test 작성
- `build.gradle`에 testcontainers 의존성 추가
- test 용 redis container 설정 추가된 `AbstractContainersSupport` 추상 클래스 추가

### 🫱🏻‍🫲🏻 To Reviewers
- test 에서 redis를 사용해야할 경우 `AbstractContainersSupport`를 `extends`해서 사용하면 됩니다!

This close #12

jo0oy committed:
#12 feat: 별점 등록 Request/Response DTO 클래스 추가

- 별점 등록시 사용될 DTO 클래스 추가
   - StarRegisterRequestDto.class
   - StarRegisterResponseDto.class

jo0oy committed:
#12 feat: 별점 등록 비즈니스 로직 구현

- StarService
   - 별점 등록 메서드 register() 구현
- 별점 등록시 필요한 회원/상품 엔티티 존재하지 않을 때 던져줄 Custom Exception 클래스 추가
   - EntityNotFoundException.class

jo0oy committed:
#12 feat: 별점 등록 API 구현

- StarRestController
   - 별점 등록 API
   - POST /api/stars

jo0oy committed:
#12 feat: 별점 도메인 관련 RestControllerAdvice 구현

jo0oy committed:
#12 test: test 프로파일에 spring.jpa.show-sql 설정 추가

jo0oy committed:
#12 test: 별점 등록 기능 Service/Controller Test 코드 작성

jo0oy committed:
#12 test: 별점 등록 API 문서화를 위한 REST Docs 테스트 추가

jo0oy committed:
#12 test: testcontainers 의존성 추가 및 redis test container 설정 추상클래스 추가

- build.gradle
   - testImplementation 'org.springframework.boot:spring-boot-testcontainers'
- AbstractContainersSupport.class
   - redis test container 설정 클래스를 추상 클래스로 구현
   - 'extends' 사용해 손쉽게 redis test container 설정을 재사용할 수 있도록 구현

jo0oy committed:
Merge branch 'develop' into feature/star-register

wocjf0513 commented on src/main/java/com/fc/shimpyo_be/domain/star/exception/EntityNotFoundException.java:
global controll advice에서 applicationException을 catch하고 있기 때문에  RuntimeException을 extends하는 ApplicationException을 extends하는 방향으로 통일하면 좋을거같아요~ 💯

jo0oy replied:
넵 수정하겠습니다!

wocjf0513 commented on src/test/java/com/fc/shimpyo_be/config/AbstractContainersSupport.java:
감사합니다. 잘 사용하겠습니다 👍

JeongUijeong commented on src/main/java/com/fc/shimpyo_be/domain/star/controller/StarRestController.java:
```diff
@@ -0,0 +1,38 @@
+    private final StarService starService;
+
+    @PostMapping
+    public ResponseEntity<?> register(@Valid @RequestBody StarRegisterRequestDto request) {
```
ResponseEntity<?> 처럼 제네릭을 쓰는 대신 ResponseEntity<ResponseDto<StarResponseDto>> 를 쓰면 더 명시적일 것 같아요~ㅎㅎ

JeongUijeong commented on src/main/java/com/fc/shimpyo_be/domain/star/dto/request/StarRegisterRequestDto.java:
```diff
@@ -0,0 +1,18 @@
+
+public record StarRegisterRequestDto(
+    @NotNull(message = "별점 등록 회원 아이디는 필수값입니다.")
+    Long memberId,
```
회원 아이디는 요청 필드로 받는 것보다 SecurityUtil의 getCurrentMemberId 메서드를 이용하는 것은 어떤가요?

jo0oy replied:
넵넵 알겠습니다!

JeongUijeong commented on src/test/java/com/fc/shimpyo_be/config/AbstractContainersSupport.java:
감사합니다!!! 😁

JeongUijeong commented on src/test/java/com/fc/shimpyo_be/domain/star/unit/service/StarServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+@SpringBootTest
```
서비스 테스트는 단위 테스트로 진행해주시면 좋을 것 같아요! 컨트롤러 같은 경우는 Security 설정을 적용하기 위해 제한적으로 @SpringBootTest를 사용한 통합테스트를 진행하고 있습니다만, 서비스는 Security 설정이 필요없으니 단위테스트로 작성해주시면 감사하겠습니다 😊

jo0oy replied:
옙 알겠습니다 :) ㅎㅎㅎ 앞으로 단위 테스트로 진행할게요!

JeongUijeong reviewed (changes requested):
수고 많으셨습니다! 코드 너무 깔끔하고 좋네요~ ㅎㅎ 몇 가지 코멘트 달아 뒀으니, 확인해주시고, 수정 부탁드립니다! 🩷

jo0oy committed:
Merge branch 'develop' into feature/star-register

jo0oy committed:
#12 fix: exception 처리 수정 및 별점 등록시 memberId 시큐리티 인증 객체에서 가져오도록 비즈니스 로직 수정

- 추상 Exception 클래스인 ApplicationException을 상속한 Custom Exception을 던질 수 있도록 구현
- RequestDto로 받던 memberId를 현재 인증 객체에서 꺼내와서 별점 등록시 사용할 수 있도록 구현
- 위의 수정에 따른 Test 코드 수정 업데이트

jo0oy reviewed (commented):
여러분 코드리뷰 감사해용~🙂

jo0oy committed:
#12 fix: ResponseEntity<?>를 명시적인 타입으로 수정

jo0oy committed:
#12 fix: StarServiceTest @SpringBootTest에서 Mockito Test로 수정

JeongUijeong reviewed (approved):
수고하셨습니다 ㅎㅎ 최고!!
